### PR TITLE
Added CCR rolling upgrade tests

### PR DESCRIPTION
**Pull request by martijnvg**
Tests index following and auto following in a rolling upgrade scenario. 

Test are written as Java tests using low level client instead of yaml tests, because CCR needs assertBusy(...) snippets to be tested correctly and that is not possible in yaml tests.

### CODE REVIEW

**Comment by martijnvg** [comment]
This test is currently marked as WIP as it depends on a bug fix: #36647 and
sometime it fails with serialization errors that occur when a 6.6.0-SNAPSHOT node receives a shard changes reponse from a 7.0.0-SNAPSHOT node:

```
Caused by: java.lang.IllegalArgumentException: No version type match [-1]
        at org.elasticsearch.index.VersionType.fromValue(VersionType.java:368) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Index.<init>(Translog.java:1151) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Index.<init>(Translog.java:1120) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Operation.readOperation(Translog.java:1072) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.common.io.stream.StreamInput.readArray(StreamInput.java:734) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.xpack.ccr.action.ShardChangesAction$Response.readFrom(ShardChangesAction.java:282) ~[?:?]
```

I'm not yet sure how this can happen, because the serialization logic in `Translog.Index` looks okay to me in both master and 6.x branches.

**Comment by martijnvg** [comment]
I've also noticed this serialization error in bulk shards operations api:

```
[2018-12-14T19:52:59,864][WARN ][o.e.x.c.a.ShardFollowNodeTask] [upgraded-node-0] shard follow task encounter non-retryable error
org.elasticsearch.transport.RemoteTransportException: [node-2][127.0.0.1:34723][indices:data/write/bulk_shard_operations[s]]
Caused by: java.lang.IllegalArgumentException: No version type match [-1]
        at org.elasticsearch.index.VersionType.fromValue(VersionType.java:368) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Index.<init>(Translog.java:1151) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Index.<init>(Translog.java:1120) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.index.translog.Translog$Operation.readOperation(Translog.java:1072) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.common.io.stream.StreamInput.readCollection(StreamInput.java:959) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.common.io.stream.StreamInput.readList(StreamInput.java:941) ~[elasticsearch-7.0.0-SNAPSHOT.jar:7.0.0-SNAPSHOT]
        at org.elasticsearch.xpack.ccr.action.bulk.BulkShardOperationsRequest.readFrom(BulkShardOperationsRequest.java:54) ~[?:?]

```

**Comment by dnhatn** [comment]
@martijnvg I have merged #36676 which should fix the translog serialization issue.

**Comment by martijnvg** [comment]
run the gradle build tests 1

**Comment by martijnvg** [comment]
run the gradle build tests 2

**Comment by martijnvg** [comment]
@dnhatn I've update this PR. Do you want to take another look?

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        assertOK(client().performRequest(new Request("POST", "/" + followIndex + "/_ccr/pause_follow")));
+    }
+
+    private static void unfollow(String followIndex) throws IOException {
+        assertOK(client().performRequest(new Request("POST", "/" + followIndex + "/_ccr/unfollow")));
+    }
+
+    private static void putAutoFollowPattern(String patternName) throws IOException {
+        Request request = new Request("PUT", "/_ccr/auto_follow/" + patternName);
+        request.setJsonEntity("{\"leader_index_patterns\": [\"logs-*\"], \"remote_cluster\": \"local\"," +
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: It's easier to follow if we pass the pattern (that's `logs-*`) as a parameter.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+    private void assertNumberOfSuccessfulFollowedIndices(int expectedNumberOfSuccessfulFollowedIndices) throws IOException {
+        Request statsRequest = new Request("GET", "/_ccr/stats");
+        Map<?, ?> response = toMap(client().performRequest(statsRequest));
+        LOGGER.info("AUTO FOLLOW STATS={}", response.get("auto_follow_stats"));
+        Integer actualSuccessfulFollowedIndices = ObjectPath.eval("auto_follow_stats.number_of_successful_follow_indices", response);
+        assertThat(actualSuccessfulFollowedIndices, equalTo(expectedNumberOfSuccessfulFollowedIndices));
+    }
+
+    private void assertFollowerGlobalCheckpoint(String followerIndex, int expectedFollowerCheckpoint) throws IOException {
+        Request statsRequest = new Request("GET", "/" + followerIndex + "/_ccr/stats");
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: I think a seq_no stats of the following index would give us a stronger confidence than the ccr stats.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        Map<?, ?> response = toMap(client().performRequest(statsRequest));
+        Integer actualSuccessfulFollowedIndices = ObjectPath.eval("auto_follow_stats.number_of_successful_follow_indices", response);
+        if (actualSuccessfulFollowedIndices != null) {
+            return actualSuccessfulFollowedIndices;
+        } else {
+            return -1;
+        }
+    }
+
+    private void assertNumberOfSuccessfulFollowedIndices(int expectedNumberOfSuccessfulFollowedIndices) throws IOException {
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: Maybe remove this method and assert the result of "getNumberOfSuccessfulFollowedIndices"?

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+                deleteAutoFollowPattern("test_pattern");
+
+                pauseFollow("copy-" + leaderIndex1);
+                closeIndex("copy-" + leaderIndex1);
+                unfollow("copy-" + leaderIndex1);
+
+                pauseFollow("copy-" + leaderIndex2);
+                closeIndex("copy-" + leaderIndex2);
+                unfollow("copy-" + leaderIndex2);
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: Maybe do all pause, close and unfollow in a single method?

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+                assertDocumentExists(leaderIndex, "1");
+                assertBusy(() -> {
+                    assertFollowerGlobalCheckpoint(followerIndex, 0);
+                    assertDocumentExists(followerIndex, "1");
+                });
+                break;
+            case MIXED:
+                if (SECOND_ROUND == false) {
+                    index(leaderIndex, "2");
+                    assertDocumentExists(leaderIndex, "2");
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: Maybe assert the existence of both doc1 and doc2? You can extend `assertDocumentExists` to accept varargs.

**Review by dnhatn — COMMENTED** [review]
Thanks @martijnvg. I like these tests - they are easy to follow. I left some minor comments.

**Comment by martijnvg** [comment]
@dnhatn I've updated the PR.

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                        assertDocumentExists(followerIndex, "3");
+                    });
+                }
+                break;
+            case UPGRADED:
+                index(leaderIndex, "4");
+                assertDocumentExists(leaderIndex, "4");
+                assertBusy(() -> {
+                    assertFollowerGlobalCheckpoint(followerIndex, 3);
+                    assertDocumentExists(followerIndex, "4");
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: nit: can we also verify doc1, doc2, and doc3?

File: x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                    assertBusy(() -> {
+                        assertFollowerGlobalCheckpoint(followerIndex, 1);
+                        assertDocumentExists(followerIndex, "1", "2");
+                    });
+                } else {
+                    index(leaderIndex, "3");
+                    assertDocumentExists(leaderIndex, "3");
+                    assertBusy(() -> {
+                        assertFollowerGlobalCheckpoint(followerIndex, 2);
+                        assertDocumentExists(followerIndex, "3");
```
Comment (dnhatn) [inline x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/CCRIT.java]: nit: can we also verify doc1, and doc2

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks @martijnvg. I left minor nits but feel free to merge once CI is happy.

**Comment by martijnvg** [comment]
I will backport this later to 6.x, if I didn't encounter build failures related to this change.

**Pull request merged**